apps/cpp/avifgainmaputil.cc

```diff
@@ -0,0 +1,386 @@
+                << " " << unknown_options[0];
+      for (size_t i = 1; i < unknown_options.size(); ++i) {
+        std::cerr << ", " << unknown_options[i];
+      }
+      std::cerr << "\n";
+      return AVIF_RESULT_INVALID_ARGUMENT;
+    }
+    return AVIF_RESULT_OK;
+  }
+  // Runs the command.
+  virtual avifResult Run() = 0;
+
+  // Returns the name of the command that should be used to invoke it on the
+  // command line.
+  virtual std::string GetName() = 0;
+  // Returns a one line description of what the command does.
+  virtual std::string GetDescription() = 0;
+  // Returns example parameters to invoke this command.
+  virtual std::string GetUsageParams() = 0;
+  // Returns a full description of how to use this command.
+  std::string GetUsage() {
+    return "Usage: avifgainmaputil " + GetName() + " " + GetUsageParams() +
+           "\n\n" + options_.help();
+  }
+
+ protected:
+  popl::OptionParser options_;
+};
+
+class HelpCommand : public ProgramCommand {
+ public:
+  std::string GetName() override { return "help"; }
+  std::string GetDescription() override { return "Prints a command's usage"; }
+  std::string GetUsageParams() override { return "<command>"; }
+  avifResult Run() override {
+    // Actual implementation is in the main function because it needs access to
+    // the list of commands.
+    return AVIF_RESULT_OK;
+  }
+};
+
+template <typename T>
+inline T Clamp(T x, T low, T high) {  // Only exists in C++17.
+  return (x < low) ? low : (high < x) ? high : x;
+}
+
+// Writes an image in any of the supported formats based on the file extension.
+avifResult WriteImage(const avifImage* image, std::string output_filename,
+                      int quality, int speed) {
+  quality = Clamp(quality, 0, 100);
+  speed = Clamp(quality, 0, 10);
+  const avifAppFileFormat output_format =
+      avifGuessFileFormat(output_filename.c_str());
+  if (output_format == AVIF_APP_FILE_FORMAT_UNKNOWN) {
+    std::cerr << "Cannot determine output file extension: " << output_filename
+              << "\n";
+    return AVIF_RESULT_INVALID_ARGUMENT;
+  } else if (output_format == AVIF_APP_FILE_FORMAT_Y4M) {
+    if (!y4mWrite(output_filename.c_str(), image)) {
+      return AVIF_RESULT_UNKNOWN_ERROR;
+    }
+  } else if (output_format == AVIF_APP_FILE_FORMAT_JPEG) {
+    if (!avifJPEGWrite(output_filename.c_str(), image, quality,
+                       AVIF_CHROMA_UPSAMPLING_AUTOMATIC)) {
+      return AVIF_RESULT_UNKNOWN_ERROR;
+    }
+  } else if (output_format == AVIF_APP_FILE_FORMAT_PNG) {
+    const int compresion_level = Clamp(10 - speed, 0, 9);
+    if (!avifPNGWrite(output_filename.c_str(), image, image->depth,
+                      AVIF_CHROMA_UPSAMPLING_AUTOMATIC, compresion_level)) {
+      return AVIF_RESULT_UNKNOWN_ERROR;
+    }
+  } else if (output_format == AVIF_APP_FILE_FORMAT_AVIF) {
+    std::unique_ptr<avifEncoder, decltype(&avifEncoderDestroy)> encoder(
```

Comment:
maybe wait for https://github.com/AOMediaCodec/libavif/pull/1733 to get EncoderPtr